### ISSUE TITLE
Solved: Communication problem with Trakt

### ISSUE THREAD

**Issue by burekas7** [elgatito/plugin.video.elementum#534]
The solution:
https://github.com/elgatito/plugin.video.elementum/issues/534#issuecomment-571752668
---

I'm getting this error when trying access to the Trakt menus

Also when trying to authorize:
Unable to get Trakt code: 403
or
Internal server Error

I guess it something with Trakt api,
But the site is online, so it's strange.

Kodi Log:
https://paste.ubuntu.com/p/r4KsQV4VVJ

## Your Environment
Elementum 0.1.44
Kodi 17.6
Win10, 64 bit

**Comment by ivanich** [comment]
Experiencing the same 403 issue with kodi-18.5 on libreelec.

**Comment by elgatito** [comment]
Trakt has disabled current Elementum API key due to big amount of plays in the history of connected accounts.

**As long as this addon is not supported anymore, there is no planned date of adding new API key (with need of re-authorizing).**

**Comment by burekas7** [comment]
> Experiencing the same 403 issue with kodi-18.5 on libreelec.
> Looks like default elementum trakt api keys were banned/removed, so I just created my own trakt api keys and added them to .kodi/userdata/addon_data/plugin.video.elementum/settings.xml and now elementum works for me.

How do I create my own API Key?
I go to the API section in the Trakt site, but what should I need to do next?

And what is the difference between API Client ID and API Client Secret in the settings?

**Comment by lazymind** [comment]
> Trakt has disabled current Elementum API key due to big amount of plays in the history of connected accounts.
> 
> **As long as this addon is not supported anymore, there is no planned date of adding new API key (with need of re-authorizing).**

Please save trakt in elementum...
It works well with own (new) API key.
Let us use it with own API key.
Thank you.

**Comment by burekas7** [comment]
@lazymind 
He didn't say that, he talked about the api key and not the Trakt in the addon.
This is the best addon regarding to Trakt support.

**Comment by burekas7** [comment]
### **How to Create Your Own Trakt API Key**

1. Click here:
https://trakt.tv/oauth/applications/new
To go to the Trakt API application signup form. You’ll have to sign in to your Trakt account, so create one if you don’t have one yet and then make sure you are redirected to a page with the heading “Create a Trakt.tv API App”.

2. **Fill out the form as follows:**
2.1 Give your API a name.
2.2 Don’t worry about entering in an icon.
2.3 Give your API a description.
2.4 Enter in a redirection URL of Google.com
2.5 Leave the Javascript origins blank
2.6 Make sure /checkin and /scrobble are checked
2.7 Click the Save Button

3. Copy and paste your Client ID and Client Secret and keep them safe.

**Comment by yakovuvarov** [comment]
settings.xml file of the Android version doesn't seem to have entries for Trakt API keys in it, and I can't find them in the interface as well. Any ideas?

**Comment by burekas7** [comment]
> settings.xml file of the Android version doesn't seem to have entries for Trakt API keys in it, and I can't find them in the interface as well. Any ideas?

In Kodi, go to Elementum Settings, Trakt tab, there are two fields:
"Custom Trakt API Client ID"
"Custom Trakt API Client Secret"

Follow my instructions in my comment above.
Don't forget to save the settings.

**Comment by yakovuvarov** [comment]
Either I'm on an outdated version of Kodi (although there were no updates in the repo in the recent days, and none right now), or I'm totally blind, because there's no such fields. If the version I'm using is outdated (it is 0.1.41), then how do I update?

**Comment by burekas7** [comment]
@yakovuvarov 
The last release:
https://github.com/elgatito/plugin.video.elementum/releases/download/v0.1.44/plugin.video.elementum-0.1.44.zip

But if you are installing the Elementum repo should be installed also, and you can always leep it update through the Kodi -> Addons section if it doesn't update automatically.

**Comment by yakovuvarov** [comment]
Yep, and it didn't see any update, I had to update it from zip manually. A bit tricky, given that I'm running a set-top box setup, but doable.

And indeed, I was using an outdated version, options for custom API keys for Trakt were available since 0.1.43. Still in doubt why there were no update from the repo, though.

**Comment by antonsoroko** [comment]
thanks for the info!
2.4 for `Redirect uri:` we can "Use `urn:ietf:wg:oauth:2.0:oob` for Device authentication." as "Create a Trakt API app" page says and as it was in elementum code.

4th step is mandatory.

**Comment by burekas7** [comment]
@antonsoroko 
Thanks.

What is the advantage of using:
`urn:ietf:wg:oauth:2.0:oob`
over:
`http://www.google.com`

Because it seems that the keys I got are working well for now.

**Comment by antonsoroko** [comment]
@burekas7 

for website that wants to authorize via trakt it would be like this:

https://trakt.docs.apiary.io/#reference/authentication-oauth/authorize-application
```
Trakt redirects back to your site. If the user accepts your request, Trakt redirects back to your site with a temporary code in a code GET parameter as well as the state (if provided) in the previous step in a state parameter. If the states don’t match, the request has been created by a third party and the process should be aborted.
```

so trakt would generate redirect http request as an answer to authorization request that was generated by some website.

but in elementum code `redirect_uri` is used only for RefreshToken (Authentication is done differently, via https://trakt.docs.apiary.io/#reference/authentication-devices).
https://trakt.docs.apiary.io/#reference/authentication-oauth/get-token/exchange-refresh_token-for-access_token
and `redirect_uri` is just ignored as i can see.
https://github.com/elgatito/elementum/blob/master/trakt/trakt.go#L819

so in a nutshell `urn:ietf:wg:oauth:2.0:oob` in our case is just "not misleading" value for  `redirect_uri` .

**Comment by elgatito** [comment]
Version 0.1.45 has new default Trakt Client key. 
Anyone can create own keys as well.

**Comment by sapiora** [comment]
I apologize writing again but the problem has happened again for me. I did all the steps in the Solved comment, however there is no change. I am using Elementum 0.1.47 on Sony Android TV.
1. I have created APIs.
2. I manually wrote the APIs in Settings menu of Elementum.
3. When I clicked on Authorize button I have received Internel Server Error.

I must say that I love the application and do not want to switch to another.

**Comment by antonsoroko** [comment]
@sapiora 
Try to look into log file
http://elementum.surge.sh/contribute/log/

Maybe you missed some characters during typing the keys. Also did you saved settings before click "authorize"?
Btw, with version 47 you can actually clear keys so elementum will use its own keys.

**Comment by sapiora** [comment]
I triple checked the APIs and still get this messages "Unable to get trakt code: 403" and "Internal Server Error".

I cleared the APIs and it is the same. Will try to create new APIs.

**Comment by antonsoroko** [comment]
@sapiora
Try to look into log file, it might show the root cause
http://elementum.surge.sh/contribute/log/

Or paste it here.

**Comment by sapiora** [comment]
I decided to listen to you. Cleared the APIs, restarted KODI and the TV and I was able to authorize Elementum to trakt. Thank you very much!

The LOG files: I uploaded them thru Elementum Settings.

**Comment by gudge25** [comment]
Hi there 
i cant copy and paste API keys to kodi 
is there  a way to make this using config files?

**Comment by rojinegr0** [comment]
@gudge25 use the kore app

**Comment by borzaka** [comment]
Update here your API keys:
`\\Libreelec\Userdata\addon_data\plugin.video.elementum\settings.xml`

**Comment by gudge25** [comment]
> Update here your API keys:
> `\\Libreelec\Userdata\addon_data\plugin.video.elementum\settings.xml`
> 
> The list are back (trending Movies/TV Shows), but no cover for them :( Another API issue?

Thanks
